src/libraries/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/CertificateRequestUsageTests.cs

```diff
@@ -552,6 +552,42 @@ public static void InvalidSignatureAlgorithmEncoding(string caseName, string sig
 #endif
         }
 
+        [Fact]
+        public static void ECDSA_Signing_ECDH()
+        {
+            using ECDsa ecdsa = ECDsa.Create(ECCurve.NamedCurves.nistP256);
+            using ECDiffieHellman ecdh = ECDiffieHellman.Create(ECCurve.NamedCurves.nistP256);
+
+            CertificateRequest issuerRequest = new CertificateRequest(
+                new X500DistinguishedName("CN=root"),
+                ecdsa,
+                HashAlgorithmName.SHA256);
+
+                issuerRequest.CertificateExtensions.Add(
+                    new X509BasicConstraintsExtension(true, false, 0, true));
+
+            CertificateRequest request = new CertificateRequest(
+                new X500DistinguishedName("CN=test"),
+                new PublicKey(ecdh),
+                HashAlgorithmName.SHA256);
+
+            request.CertificateExtensions.Add(
+                new X509BasicConstraintsExtension(false, false, 0, true));
+            request.CertificateExtensions.Add(
+                new X509KeyUsageExtension(X509KeyUsageFlags.KeyAgreement, true));
+
+            DateTimeOffset notBefore = DateTimeOffset.UtcNow;
+            DateTimeOffset notAfter = notBefore.AddDays(30);
+            byte[] serial = new byte[] { 1, 2, 3, 4, 5, 6, 7, 8 };
+
+            using X509Certificate2 issuer = issuerRequest.CreateSelfSigned(notBefore, notAfter);
+            using X509Certificate2 cert = request.Create(issuer, notBefore, notAfter, serial);
+
+            using ECDiffieHellman publicCertKey = cert.GetECDiffieHellmanPublicKey();
+            Assert.NotNull(publicCertKey);
+            Assert.Equal(ecdh.ExportSubjectPublicKeyInfo(), publicCertKey.ExportSubjectPublicKeyInfo());
+        }
+
```

Comment:
Do you see value any of these extra tests:

* new CertificateRequest(ecdsa, ...) with KeyAgreement KU makes for a valid ECDH/invalid ECDSA cert
* new CertificateRequest(new PublicKey(ecdh), ...) without KeyAgreement makes the cert valid for ECDSA and ECDH
* Other sorts of "flexibility has increased" / "the output key can't be used to rerun this test".

The first bullet is actually interesting in that CreateSelfSigned will... probably... fail? Since CopyWithPrivateKey... should... fail?

Reply:
> CreateSelfSigned will... probably... fail? Since CopyWithPrivateKey... should... fail?

Yeah. It fails with

> The provided key does not match the public key algorithm for this certificate.

This is already existing behavior. I think the behavior is... reasonable, but maybe could be something we have additional validation on in `CreateSelfSigned` so the exception is clearer. If anything it seems not quite right that the result is an `ArgumentException`. I don't think folks should reasonably expect to create a cert with an ECDSA cert and get an ECDH back out.

Reply:
Actually we throw an `ArgumentException` in a lot of other cases for `Create`, so maybe it's closer to "okay" than I thought.

Reply:
> I don't think folks should reasonably expect to create a cert with an ECDSA cert and get an ECDH back out.

I agree, but if it works now (especially if it works in already shipped versions) we probably want to keep it working.

> but maybe could be something we have additional validation on in CreateSelfSigned so the exception is clearer.

I'm amenable to that (but it probably makes more sense as a followup change).

> Actually we throw an ArgumentException in a lot of other cases for Create, so maybe it's closer to "okay" than I thought.

We could (in a change other than this one) try to clean those up, too.

Reply:
>  if it works now (especially if it works in already shipped versions)

It does not nor did it. Opened #42751 to improve it later.

Reply:
Well, the thing I'm concerned works and could potentially not in the future is Create, not CreateSelfSigned.  (But it'd require us saving a lot more state than we currently do, so it's not super likely to break)